packages/legend-query/src/components/QueryBuilderCustomDatePicker.tsx

```diff
@@ -0,0 +1,1267 @@
+  ONE_WEEK_AGO = 'One Week Ago',
+  CUSTOM_DATE = 'Custom Date',
+  PREVIOUS_DAY_OF_WEEK = 'Previous ... of Week',
+  FIRST_DAY_OF = 'First day of...',
+  LATEST_DATE = '%latest',
+}
+
+enum QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT {
+  DAYS = 'Day(s)',
+  WEEKS = 'Week(s)',
+  MONTHS = 'Month(s)',
+  YEARS = 'Year(s)',
+}
+
+enum QUERY_BUILDER_FIRST_DAY_OF_UNIT {
+  WEEK = 'Week',
+  MONTH = 'Month',
+  QUARTER = 'Quarter',
+  YEAR = 'Year',
+}
+
+enum QUERY_BUILDER_DAY_OF_WEEK {
+  MONDAY = 'Monday',
+  TUESDAY = 'Tuesday',
+  WENDNESDAY = 'Wednesday',
+  THURSDAY = 'Thursday',
+  FRIDAY = 'Friday',
+  SATURDAY = 'Saturday',
+  SUNDAY = 'Sunday',
+}
+
+enum QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION {
+  BEFORE = 'Before',
+  AFTER = 'After',
+}
+
+enum QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT {
+  TODAY = 'Today',
+  NOW = 'Now',
+  FIRST_DAY_OF_YEAR = 'Start of Year',
+  FIRST_DAY_OF_QUARTER = 'Start of Quarter',
+  FIRST_DAY_OF_MONTH = 'Start of Month',
+  FIRST_DAY_OF_WEEK = 'Start of Week',
+}
+
+/**
+ * DatePickerOption is the base class being used to display and generate the corresponding pure date function.
+ * @label is the text that shows up in the valueSpecification box.
+ * @value is the selected value in date-dropdown.
+ */
+class DatePickerOption {
+  label: string;
+  value: string;
+
+  constructor(label: string, value: string) {
+    this.label = label;
+    this.value = value;
+  }
+}
+
+class CustomDateOption extends DatePickerOption {
+  duration: number;
+  unit: QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT | undefined;
+  direction: QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION | undefined;
+  referenceMoment:
+    | QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT
+    | undefined;
+
+  constructor(
+    label: string,
+    value: string,
+    duration: number,
+    unit: QUERY_BUILDER_CUSTOM_DATE_OPTION_UNIT | undefined,
+    direction: QUERY_BUILDER_CUSTOM_DATE_OPTION_DIRECTION | undefined,
+    referenceMoment:
+      | QUERY_BUILDER_CUSTOM_DATE_OPTION_REFERENCE_MOMENT
+      | undefined,
+  ) {
+    super(label, value);
+    this.duration = duration;
+    this.unit = unit;
+    this.direction = direction;
+    this.referenceMoment = referenceMoment;
+  }
+
+  generateDisplayableLabel(): string {
+    return [
+      this.duration,
+      this.unit,
+      this.direction,
+      this.referenceMoment,
+    ].join(' ');
+  }
+
+  updateLabel(): void {
+    this.label = this.generateDisplayableLabel();
+  }
+}
+
+class CustomFirstDayOfOption extends DatePickerOption {
+  unit: QUERY_BUILDER_FIRST_DAY_OF_UNIT | undefined;
+
+  constructor(
+    label: string,
+    value: string,
+    unit: QUERY_BUILDER_FIRST_DAY_OF_UNIT | undefined,
+  ) {
+    super(label, value);
+    this.unit = unit;
+  }
+}
+
+class CustomPreviousDayOfWeekOption extends DatePickerOption {
+  day: QUERY_BUILDER_DAY_OF_WEEK;
+
+  constructor(label: string, value: string, day: QUERY_BUILDER_DAY_OF_WEEK) {
+    super(label, value);
```

Comment:
I don't think we need to pass in the value here (maybe the label?) since it is always the same